### PR TITLE
Improve redraw performance

### PR DESCRIPTION
**Pull request by philplckthun**
Improved redraw performance by removing unnecessary clipping.

Sorry for the messy commits, but it took me a while to grasp which parts of the code were doing what.

### CODE REVIEW

**Comment by philplckthun** [comment]
This essentially fixes #105 by changing as few code as possible.

**Comment by philplckthun** [comment]
@rogual I can remove the editorconfig, no problem, but I really advise you to keep it, as it's very valuable in a growing project. I will squash the commits down today or tomorrow, no problem, but technically it's unnecessary, as it's just effectively erasing a continuous git history ;)

**Comment by rogual** [comment]
Ah, didn't realize editorconfig was there on purpose. Had a look at their website; seems interesting. I'm not necessarily against using this but it doesn't belong in this change.

**Comment by philplckthun** [comment]
@rogual Squashed down the unnecessary commits. I'll add an issue for the editorconfig to discuss it. ;)

**Comment by gglanzani** [comment]
I'm trying the last neovim with this branch merged but my leader (,) stopped working. This does not happen if I try neovim "naked".

Has this something to do with this PR?

**Comment by rogual** [comment]
@philplckthun thanks for squashing this down. Good call on leaving as 2 commits, that's much clearer than it would be as 1.

@gglanzani Can reproduce, but also on master, so the problem isn't in this commit. Some of my other key-sequence bindings are broken too, like anythings starting with [ or ].

**Comment by rogual** [comment]
@gglanzani I've opened an issue for this https://github.com/rogual/neovim-dot-app/issues/112

**Comment by philplckthun** [comment]
@rogual This is essentially ready to merge now. The latest commit is another significant boost as it gets rid of the whole "preparation loop" that builds up a vector-array. Furthermore the background is now drawn in a single step instead of separately per character.

**Comment by philplckthun** [comment]
@rogual Also limited the drawing area as some unicode was overlapping the line height. This can easily be fixed by correcting the font height formula, which is not quite the same as iTerm's, I guess.

But for now I readded the clipping, which is now just executed once. Shouldn't be too bad, but I think we should remove it as soon as possible.

**Comment by philplckthun** [comment]
@rogual Newest commit renders groups of characters together. It occurred to me that this actually is not a bad idea, given that we're only strictly using monospace fonts.

This further improves redraw performance (significantly I believe) and fixes #107 additionally.

We should merge this soon ;)

**Comment by gglanzani** [comment]
I'm trying it right now, but with large files (largish: 1000 lines) with lots of folds (20-25), pressing j repeatedly still lags behind (i.e. I stop pressing j, and the UI keeps going down a bit).

Is this an unrelated problem?

**Comment by philplckthun** [comment]
@gglanzani I think there's probably not much I can do about that. Just opened a 50,000 lines file and it was rendered fine and the scrolling didn't lag. It was without any syntax highlighting though.

Can you film a small video of your lags maybe? Can you also try to load Neovim.app without your vimrc? Maybe it's a plugin of yours or something? (no offense :wink:)

File: src/redraw.mm [hunk 1]
```diff
@@ -70,123 +70,65 @@ - (void) doAction:(RedrawCode::Enum)code withItem:(const object &)item_o
...
+        // Draw background separately [width = item_sz]
+        NSRect bgrect = [self
+            viewRectFromCellRect:CGRectMake(mCursorPos.x, mCursorPos.y, item_sz - 1, 1)];
+        [bg set];
+        NSRectFill(bgrect);
 
-        static std::string run;
-        run.clear();
+        // Init string
+        NSString *nsrun = @"\u202d";
```
Comment (rogual) [inline src/redraw.mm]: Please add a comment explaining the meaning of \u202d. (It used to be near the `// Force left-to-right rendering` comment)

File: src/redraw.mm [hunk 2]
```diff
@@ -70,123 +70,65 @@ - (void) doAction:(RedrawCode::Enum)code withItem:(const object &)item_o
 {
     int item_sz = item_o.via.array.size;
     object *arglists = item_o.via.array.ptr + 1;
-    int narglists = item_sz - 1;
-
 
     if (code == RedrawCode::put) {
+        // Separate Front- and Backend colors
```
Comment (rogual) [inline src/redraw.mm]: s/Backend/Background

Comment (philplckthun) [inline src/redraw.mm]: sorry. Web development keeps spoiling me

File: src/redraw.mm [hunk 3]
```diff
@@ -70,123 +70,65 @@ - (void) doAction:(RedrawCode::Enum)code withItem:(const object &)item_o
...
+        NSColor *fg = [mTextAttrs objectForKey:NSForegroundColorAttributeName];
+        NSColor *bg = [mTextAttrs objectForKey:NSBackgroundColorAttributeName];
 
-        /* This loop is getting a bit messy; feel free to rewrite. But, these
-           lines should render properly and be aligned:
-
-            iiiiiiiiiiiiiiiiiiii|
-            MMMMMMMMMMMMMMMMMMMM|
-            サパラサパサパラサパ|
-            བཔོལག་བཔོལག་བཔོལག་བཔོལག་|
```
Comment (rogual) [inline src/redraw.mm]: Please don't remove this alignment test block.

File: src/redraw.mm [hunk 4]
```diff
@@ -70,123 +70,65 @@ - (void) doAction:(RedrawCode::Enum)code withItem:(const object &)item_o
...
-            int sz = lens[i];
+        // Force left-to-right rendering
+        nsrun = [nsrun stringByAppendingString:@"\u202c"];
 
-            if (sz == 0)
-                continue;
-
-            NSString *nsrun = [NSString stringWithUTF8String:run.c_str()];
+        CGContextSaveGState(mCanvasContext);
+        // TODO: Remove when font height formula is corrected (updateCharSize)
```
Comment (rogual) [inline src/redraw.mm]: Are you sure there is a problem with updateCharSize? Do you know what the problem is?

File: src/redraw.mm [hunk 5]
```diff
@@ -70,123 +70,65 @@ - (void) doAction:(RedrawCode::Enum)code withItem:(const object &)item_o
...
-
-            mCursorPos.x += sz;
-        }
+        mCursorPos.x += item_sz - 1;
 
+        CGContextRestoreGState(mCanvasContext);
         [self setNeedsDisplay:YES];
     }
-    else for (int i=0; i<narglists; i++) {
+    else for (int i = 0; i < item_sz - 1; i++) {
```
Comment (rogual) [inline src/redraw.mm]: Why `- 1`? I mean, I know why, but having the `narglists` var made this more obvious. Did you remove it on purpose?

File: src/redraw.mm [hunk 6]
```diff
@@ -360,14 +303,14 @@ - (void) doAction:(RedrawCode::Enum)code withArgc:(int)argc argv:(const object *
...
-            totalRect.origin.y += amt * mCharSize.height;
+            totalRect.origin.y = floor(totalRect.origin.y + amt * mCharSize.height);
 
             CGContextSaveGState(mCanvasContext);
             CGContextClipToRect(mCanvasContext, dest);
             drawBitmapContext(mCanvasContext, mCanvasContext, totalRect);
             CGContextRestoreGState(mCanvasContext);
 
-            /* Clear the newly-visible lines */
+            // Clear the newly-visible lines
```
Comment (rogual) [inline src/redraw.mm]: Please don't do this.

Comment (rogual) [inline src/redraw.mm]: I need to get round to specifying a style guide I guess.

**Comment by rogual** [comment]
Many thanks for all this work. I just want to be sure I undertsand the change before merging it. Could you explain why all the extra floor()s and ceil()s are necessary?

**Comment by philplckthun** [comment]
@rogual I fix the issues and add a new commit. I'll also need to fix the double-width characters

File: src/redraw.mm (hunk 4 above)
Comment (philplckthun) [inline src/redraw.mm]: I'll open a separate issue for this with details, comparisons and maybe some screenshots. But what I can already say is: Full height unicode characters are cut off a little and there is some "magic formula" in between the one we used before and the one we're using now. It's weird.

File: src/redraw.mm (hunk 5 above)
Comment (philplckthun) [inline src/redraw.mm]: Yea, I removed it as it was just used at that place. I thought it's pretty obvious. We can still add a comment about this, but I guess it's not a big deal.

File: src/redraw.mm (hunk 6 above)
Comment (philplckthun) [inline src/redraw.mm]: Yea, a brief style guide would be nice. There are plenty ones out there. Maybe you should just link to one and write up a "CONTRIBUTE.md"? :)

Just my 2 cents to the comments: If they're one liners I tend to use the "double-slash" style as it doesn't intervene with multi-line comments. Therefore it's easier to comment out big chunks of the code. Maybe you've got other reasons to always use multi-line comments. Anyway, it's a matter of preference and taste, and a styleguide is thus a great idea :+1:

**Comment by philplckthun** [comment]
@rogual This should be fine now :) Readded the comments you asked for. This commit cancels the string build if the last char is a double width character. So it separates double width characters and keeps them aligned.

I couldn't get: "བཔོལག་བཔོལག་བཔོལག་བཔོལག་" to align correctly as those are just ligatures, which compress and not expand, so that's hard to detect. I noticed that iTerm also does a poor job detecting them. I guess in the future we can look into calculating the width of all strings to be redrawn but atm that's probably too slow.

**Comment by bambu** [comment]
:+1 

So fast! Lots of asian characters are displayed nicely (not so with the original grouped commit).
As for the hindi example provided in the comment, MacVim / VimR seem to also fail to properly display it. 
Tried using FiraCode font as mentioned in #107 and the ligatures look nice. I might even start using it from now on.

**Comment by gglanzani** [comment]
@philplckthun I was pretty sure I already replied with a video, but apparently I didn't. 

Video is here https://www.youtube.com/watch?v=b3zv4n_l8HI (sorry for the bg music, I'm not so handy with video capture).

Note that after 2s, I stop pressing j, but the cursor keeps getting down. This is also present in neovim in the terminal, but it's much less laggy.

Scrolling is fine, as you can see towards the end.

**Comment by philplckthun** [comment]
@gglanzani Can you open an issue for this? Pretty sure this is unrelated. Just looks like the redraw loop is spammed with the press events. The key repeat rate is probably higher than it can be handled

File: src/redraw.mm [hunk 7]
```diff
@@ -70,123 +70,88 @@ - (void) doAction:(RedrawCode::Enum)code withItem:(const object &)item_o
...
 
     if (code == RedrawCode::put) {
+        /*
+        This is the main redraw loop, which draws consecutive characters on one line.
+        The background is drawn separately from the foreground and the text
+        attributes are set in another event.
+        These lines should render properly and be aligned:
+        iiiiiiiiiiiiiiiiiiii|
+        MMMMMMMMMMMMMMMMMMMM|
+        サパラサパサパラサパ|
```
Comment (rogual) [inline src/redraw.mm]: There were four lines in this test block; please add the other one back in.

File: src/redraw.mm (hunk 6 above)
Comment (rogual) [inline src/redraw.mm]: I'm not too fussed about what comment style you use when writing code; I just prefer for PRs not to have changes that don't relate to the issue at hand, such as changing the style of existing comments. I'll get on with that CONTRIBUTE file :)

**Comment by rogual** [comment]
> I couldn't get: "བཔོལག་བཔོལག་བཔོལག་བཔོལག་" to align correctly as those are just ligatures, which compress and not expand, so that's hard to detect. I noticed that iTerm also does a poor job detecting them. I guess in the future we can look into calculating the width of all strings to be redrawn but atm that's probably too slow.

OK, I only just realized this. I hope we don't have to sacrifice correctness for speed here. I mean, the whole reason why Neovim.app was slow in the first place is because it was getting cases like this right.

**Comment by philplckthun** [comment]
@rogual Well I hate sacrificing correctness here, but unicode and font rendering is fucked up anyway. I mean this whole issue just arises as there is no monospaced font for Chinese, Japanese, Korean, Arabian, Indish whatever on the Mac.

iTerm, MacVim and VimR don't get it right as well, so I guess it is ok to sacrifice performance here.

We can maybe detect the unicode range of those characters manually and separate them as well, but let's add an issue for that.

**Comment by rogual** [comment]
> unicode and font rendering is fucked up anyway

What do you mean? It's correct on master.

> iTerm, MacVim and VimR don't get it right as well, so I guess it is ok to make that sacrifice here.

On the other hand, Terminal.app does get it right. I'd rather emulate apps that do things right rather than ones that don't.

I'm leaning towards merging this, though, because I think more people are affected by the slowness than by certain scripts not working.

**Comment by philplckthun** [comment]
@rogual I meant unicode and font rendering in general, not Neovim.app in particular. We can still merge this usable today, add an issue for this, and I can surely get this fixed soon, when I come up with another idea on how to do this ;)

I doubt that a script will fail because of this. Even then this doesn't influence the coding itself. It is just aesthetically incorrect.

**Comment by philplckthun** [comment]
@rogual Added an issue #121 for the Tibetan characters. I think this is merge-able. I've been using it for the whole week end without problems.

**Comment by rogual** [comment]
I have too, but I've been getting some rendering bugs where characters shift around by a few pixels sometimes. Trying to narrow it down, because that isn't a very good bug report :)

**Comment by philplckthun** [comment]
Can you post a screenshot so I can check this out?

**Comment by rogual** [comment]
Of course, now that I mention it, it stops happening.

**Comment by rogual** [comment]
Alright, let's do this.

**Comment by philplckthun** [comment]
@rogual Murphy's Law :shipit:

**Pull request merged**